airbyte-webapp/src/pages/connections/CreateConnectionPage/usePreloadData.ts

```diff
@@ -0,0 +1,113 @@
+   *      param match it.
+   *   2) else if there's a `connectorId` query param, we arrived at this page via a fresh
+   *      page load. This logic was added to support the airbyte-cloud's free connector
+   *      program enrollment flow (which involves a round-trip visit to a Stripe domain,
+   *      which wipes location.state), but it's not explicitly tied to it. We explicitly
+   *      navigate to the current path (including the current query string, to avoid going
+   *      around in circles) while setting the missing `location.state.connectorId`.
+   *
+   * There are several states for each kind of connector, so let's extract a bunch of named
+   * boolean variables to try to keep the combinations readable.
+   */
+  const sourceIdInBoth = sourceIdFromLocationState && sourceIdFromSearchParams === sourceIdFromLocationState;
+  const sourceIdInNeither = !sourceIdFromLocationState && !sourceIdFromSearchParams;
+  const sourceIdInSync = sourceIdInBoth || sourceIdInNeither;
+
+  const destinationIdInBoth =
+    destinationIdFromLocationState && destinationIdFromSearchParams === destinationIdFromLocationState;
+  const destinationIdInNeither = !destinationIdFromLocationState && !destinationIdFromSearchParams;
+  const destinationIdInSync = destinationIdInBoth || destinationIdInNeither;
+
+  useEffect(() => {
+    const availableConnectorIds = { ...(sourceId ? { sourceId } : {}), ...(destinationId ? { destinationId } : {}) };
+
+    // the source conditions are the top-level, with destination conditions nested inside
+    if (sourceIdInSync && destinationIdInSync) {
+      // no further action needed here
+    } else if (sourceIdFromLocationState || destinationIdFromLocationState) {
+      // location state takes precedence if search params and location.state have to be synced
+      setSearchParams(availableConnectorIds);
+    } else if (sourceIdFromSearchParams || destinationIdFromSearchParams) {
+      // here, we have to simultaneously set matching query string and location.state
+      // values to avoid a mutually recursive infinite loop:
+      //   A: set location.state and rerender; GOTO B
+      //   B: set query param and rerender; GOTO A
+      navigate(`${location.search}`, { state: availableConnectorIds, replace: true });
+    }
+  }, [
+    destinationId,
+    destinationIdFromLocationState,
+    destinationIdFromSearchParams,
+    destinationIdInSync,
+    location.search,
+    navigate,
+    setSearchParams,
+    sourceId,
+    sourceIdFromLocationState,
+    sourceIdFromSearchParams,
+    sourceIdInSync,
+  ]);
+  const source = useGetSource(sourceId);
+
+  const sourceDefinition = useSourceDefinition(source?.sourceDefinitionId);
+
+  const destination = useGetDestination(destinationId);
+  const destinationDefinition = useDestinationDefinition(destination?.destinationDefinitionId);
+
+  return { source, sourceDefinition, destination, destinationDefinition };
```

Comment:
All this instead of a refactor 😬

Reply:
All this is baby steps towards a bigger refactoring: adding an extra consideration to the existing `usePreloadData` hook made it harder to read and understand the page component itself

Reply:
It's probably worth noting that the addition of this file is mostly just a reorganization of existing code: compare it to the large block of deleted code in `CreateConnectionPage.tsx`

Reply:
https://github.com/airbytehq/airbyte/issues/23114